### ISSUE TITLE
Compiler warning with GCC: -Wfree-nonheap-object on sbo::small_vector

### ISSUE THREAD

**Issue by al42and** [AdaptiveCpp/AdaptiveCpp#1381]
**Bug summary**

The current tip of the `develop` branch, as well as v23.10.0, raise a compiler warning about `delete` misuse.

**To Reproduce**

```console
$ git checkout v23.10.0
$ cmake .. -DCMAKE_C_COMPILER=gcc-12 -DCMAKE_CXX_COMPILER=g++-12 -DLLVM_DIR=/usr/lib/llvm-17/lib/cmake/ -DWITH_SSCP_COMPILER=OFF -DWITH_OPENCL_BACKEND=OFF -DWITH_LEVEL_ZERO_BACKEND=OFF -DWITH_CUDA_BACKEND=OFF -DWITH_ROCM_BACKEND=OFF
$ make
....
In file included from /usr/include/x86_64-linux-gnu/c++/12/bits/c++allocator.h:33,
                 from /usr/include/c++/12/bits/allocator.h:46,
                 from /usr/include/c++/12/string:41,
                 from /usr/include/c++/12/stdexcept:39,
                 from /usr/include/c++/12/system_error:41,
                 from /usr/include/c++/12/mutex:41,
                 from /home/aland/AdaptiveCpp/include/hipSYCL/runtime/data.hpp:32,
                 from /home/aland/AdaptiveCpp/src/runtime/dag_builder.cpp:29:
In member function ‘void std::__new_allocator<_Tp>::deallocate(_Tp*, size_type) [with _Tp = std::shared_ptr<hipsycl::rt::dag_node>]’,
    inlined from ‘constexpr void sbo::small_buffer_vector_allocator<T, MaxSize, NonReboundT>::deallocate(void*, size_t) [with T = std::shared_ptr<hipsycl::rt::dag_node>; long unsigned int MaxSize = 8; NonReboundT = std::shared_ptr<hipsycl::rt::dag_node>]’ at /home/aland/AdaptiveCpp/include/hipSYCL/common/sbo/small_vector.hpp:53:33,
    inlined from ‘static void std::allocator_traits< <template-parameter-1-1> >::deallocate(_Alloc&, pointer, size_type) [with _Alloc = sbo::small_buffer_vector_allocator<std::shared_ptr<hipsycl::rt::dag_node>, 8, std::shared_ptr<hipsycl::rt::dag_node> >]’ at /usr/include/c++/12/bits/alloc_traits.h:345:23,
    inlined from ‘void std::_Vector_base<_Tp, _Alloc>::_M_deallocate(pointer, std::size_t) [with _Tp = std::shared_ptr<hipsycl::rt::dag_node>; _Alloc = sbo::small_buffer_vector_allocator<std::shared_ptr<hipsycl::rt::dag_node>, 8, std::shared_ptr<hipsycl::rt::dag_node> >]’ at /usr/include/c++/12/bits/stl_vector.h:387:19,
    inlined from ‘std::_Vector_base<_Tp, _Alloc>::~_Vector_base() [with _Tp = std::shared_ptr<hipsycl::rt::dag_node>; _Alloc = sbo::small_buffer_vector_allocator<std::shared_ptr<hipsycl::rt::dag_node>, 8, std::shared_ptr<hipsycl::rt::dag_node> >]’ at /usr/include/c++/12/bits/stl_vector.h:366:15,
    inlined from ‘std::vector<_Tp, _Alloc>::~vector() [with _Tp = std::shared_ptr<hipsycl::rt::dag_node>; _Alloc = sbo::small_buffer_vector_allocator<std::shared_ptr<hipsycl::rt::dag_node>, 8, std::shared_ptr<hipsycl::rt::dag_node> >]’ at /usr/include/c++/12/bits/stl_vector.h:733:7,
    inlined from ‘void std::vector<_Tp, _Alloc>::_M_move_assign(std::vector<_Tp, _Alloc>&&, std::true_type) [with _Tp = std::shared_ptr<hipsycl::rt::dag_node>; _Alloc = sbo::small_buffer_vector_allocator<std::shared_ptr<hipsycl::rt::dag_node>, 8, std::shared_ptr<hipsycl::rt::dag_node> >]’ at /usr/include/c++/12/bits/stl_vector.h:1960:7,
    inlined from ‘void std::vector<_Tp, _Alloc>::_M_move_assign(std::vector<_Tp, _Alloc>&&, std::false_type) [with _Tp = std::shared_ptr<hipsycl::rt::dag_node>; _Alloc = sbo::small_buffer_vector_allocator<std::shared_ptr<hipsycl::rt::dag_node>, 8, std::shared_ptr<hipsycl::rt::dag_node> >]’ at /usr/include/c++/12/bits/stl_vector.h:1969:18,
    inlined from ‘std::vector<_Tp, _Alloc>& std::vector<_Tp, _Alloc>::operator=(std::vector<_Tp, _Alloc>&&) [with _Tp = std::shared_ptr<hipsycl::rt::dag_node>; _Alloc = sbo::small_buffer_vector_allocator<std::shared_ptr<hipsycl::rt::dag_node>, 8, std::shared_ptr<hipsycl::rt::dag_node> >]’ at /usr/include/c++/12/bits/stl_vector.h:766:16,
    inlined from ‘sbo::small_vector<T, N>& sbo::small_vector<T, N>::operator=(sbo::small_vector<T, N>&&) [with T = std::shared_ptr<hipsycl::rt::dag_node>; long unsigned int N = 8]’ at /home/aland/AdaptiveCpp/include/hipSYCL/common/sbo/small_vector.hpp:83:31,
    inlined from ‘hipsycl::rt::dag& hipsycl::rt::dag::operator=(hipsycl::rt::dag&&)’ at /home/aland/AdaptiveCpp/include/hipSYCL/runtime/dag.hpp:47:7,
    inlined from ‘hipsycl::rt::dag hipsycl::rt::dag_builder::finish_and_reset()’ at /home/aland/AdaptiveCpp/src/runtime/dag_builder.cpp:232:22:
/usr/include/c++/12/bits/new_allocator.h:158:33: warning: ‘void operator delete(void*, std::size_t)’ called on pointer ‘<unknown>’ with nonzero offset 40 [-Wfree-nonheap-object]
  158 |         _GLIBCXX_OPERATOR_DELETE(_GLIBCXX_SIZED_DEALLOC(__p, __n));
      |                                 ^
```

**Expected behavior**

ACpp compiles with no warnings

**Describe your setup**
* ACpp v23.10.0, recent `develop` (2d0c6b6f22fdfd5e026623ae597d1836cfa94970), first commit with small-vector optimization (fcf39553dd42db08eb8252fb70fede3286caded3).
* Describe the dependencies that AdaptiveCpp sits on top of in your setup: 
   * Clang 17.0.6 from LLVM APT repos
   * GCC 12.3.0 and 11.4.0 from standard Ubuntu repos; manually-built version of GCC 14.0.1 (70b9dd8b388668df96f457b3c6ee33fc5b35f94c) I had lying around. The behavior is the same for all versions.

**Comment by illuhad** [comment]
This is unfortunately there since the small-vector optimization was introduced, and caused by the interplay of that library with libstdc++ headers. I *think* it's not something that actually causes a problem, and most likely a false positive due to the compiler static analysis not realizing that the `operator delete()` code path is not actually taken when the stack memory is in use. So far I have not seen any problem caused by it (including with asan builds).

Since we cannot fix the root issues as the warning is raised by external libraries, we could disable the warning if you prefer, but it might also be a good idea to keep an eye out for this. I think the only reason we don't see the warning with `acpp` when compiling user code is because we use `-isystem` there.

**Comment by al42and** [comment]
That fits my own analysis, but what is suspicious to me is why only `dag::operator=(dag&&)` triggers the warning, and not any other destructor call that affects `node_list_t`. With the allocator copying etc going on, I am not fully sure that nothing fishy is happening. But could be a false-positive, yes.

UPD: avoiding the move-assignment by using `dag final_dag = std::exchange(_current_dag, {});` in `dag_builder::finish_and_reset()` makes the warning go away, but this is not backed by any thorough analysis.

**Comment by al42and** [comment]
Fixed by #1383
